api.py

```diff
@@ -388,6 +388,107 @@ def set_generic_data(dataset_id, key, data_string, **kwargs):
+
+
+def convert_samples_to_entity_set(sample_list):
+    conduce_keys = ['id', 'kind', 'time', 'point', 'path', 'polygon']
+    entities = []
+    for sample in sample_list:
+        if not isinstance(sample['time'], datetime):
+            raise TypeError('time must be a datetime object', sample['time'])
+        coordinates = []
+        if 'point' in sample:
+            coordinates = [convert_coordinates(sample['point'])]
+        if 'path' in sample:
+            if len(coordinates) > 0:
+                raise KeyError('A sample may only contain one of point, path or polygon', sample)
+            if len(sample['path']) < 2:
+                raise KeyError('paths must have at least two points')
+            for point in sample['path']:
+                coordinates.append(convert_coordinates(point))
+        if 'polygon' in sample:
+            if len(coordinates) > 0:
+                raise KeyError('A sample may only contain one of point, path or polygon', sample)
+            if len(sample['polygon']) < 3:
+                raise KeyError('polygons must have at least three points')
+            for point in sample['polygon']:
+                coordinates.append(convert_coordinates(point))
+
+        sample['_kind'] = sample['kind']
+        attribute_keys = [key for key in sample.keys() if key not in conduce_keys]
+        attributes = util.get_attributes(attribute_keys, sample)
+
+        entity = {
+            'identity': sample['id'],
+            'kind': sample['kind'],
+            'timestamp_ms': util.datetime_to_timestamp_ms(sample['time']),
+            'endtime_ms': util.datetime_to_timestamp_ms(sample['time']),
+            'path': coordinates,
+            'attrs': attributes
+        }
+
+        entities.append(entity)
+
+    return {'entities': entities}
+
+
+def ingest_samples(dataset_id, sample_list, **kwargs):
+    """
+    Upload a :ref:`sample list <conduce-entities>` to the Conduce datastore.
+
+    A convenience method that adds an list of entity samples to the Conduce datastore and waits for the job to complete. This function POSTs a sample list to Conduce and :py:func:`wait_for_job` until the ingest job completes.
+
+    Parameters
+    ----------
+    dataset_id : string
+        The UUID that identifies the dataset to modify.
+    sample_list : list
+        A list of entity samples.  See :ref:`conduce-entities` for documentation on how to build a sample list.
+
+    **kwargs:
+        See :py:func:`make_post_request`
+
+    Returns
+    -------
+    requests.Response
+        Returns an error or the final response message when the job is no longer running.
+
+    Raises
+    ------
+    requests.HTTPError
+        Requests that result in an error raise an exception with information about the failure. See :py:func:`requests.Response.raise_for_status` for more information.
+    """
+
+    if not isinstance(sample_list, list):
+        sample_list = [sample_list]
+
+    if 'id' and 'time' in sample_list[0]:
```

Comment:
this seems delicate - only the first sample needs to have id and time.  I'd put this in `convert_samples_to_entity_set` and check every sample.

Also, I'd say why the sample was invalid - I'd want to know how to fix it w/o looking for this error msg in the code & seeing why it happened

Reply:
My intention here was not to validate every sample, but rather to validate that the user passed samples and not entities, my thinking being that the user would not confuse structures within a list.  They would either be samples or entities.

However maybe I should be validating every sample and providing detailed information about missing/invalid fields.

I could do something simple now, but that seems like a follow on story.